### PR TITLE
fix (testing): better handling of sqlite connections for TestSetup(...)

### PR DESCRIPTION
**Pull request by jimlambrt**
Add new option WithTestMigrationUsingDB(..) to support special case
handling of sqlite because we may want to run the migration on an
in-memory database that isn't shared  "file::memory:" or ":memory:" so
we have to be sure that we don't open a new connection, which would
create a new in-memory db vs using the existing one already opened
in TestSetup(...) see: https://www.sqlite.org/inmemorydb.html

### CODE REVIEW

File: testing_test.go [hunk 1]
```diff
@@ -65,3 +104,14 @@ func Test_CreateDropTestTables(t *testing.T) {
 		TestCreateTables(t, db)
 	})
 }
+
+// testUser is require since we can't import dbtest as it creates a circular dep
```
Comment (johanbrandhorst) [inline testing_test.go:108]:
```suggestion
// testUser is required since we can't import dbtest as it creates a circular dep
```

**Review by johanbrandhorst — APPROVED** [review]
Probably want to get someone with more context to review as well

File: testing_test.go [hunk 2]
```diff
@@ -35,25 +39,60 @@ func Test_TestSetup(t *testing.T) {
...
 	}{
 		{
-			name:    "with-migration",
-			dialect: "sqlite",
-			opt:     []TestOption{WithTestDialect(Sqlite.String()), WithTestMigration(testMigrationFn)},
+			name: "sqlite-with-migration",
+			opt:  []TestOption{WithTestDialect(Sqlite.String()), WithTestMigration(testMigrationFn)},
+			// we can't validate this, since WithTestMigration will open a new
+			// sqlite connection which will result in a new in-memory db which
+			// will only existing during the testMigrationFn... sort of silly,
```
Comment (tmessi) [inline testing_test.go:66]: Would client test code that provides the same options here also have a similar issue with their own tests? Is this something that should not be a valid combination and should result in an error?

**Review by tmessi — APPROVED** [review]
Looks good, just one question

File: testing.go [hunk 3]
```diff
@@ -97,11 +98,43 @@ func TestSetup(t *testing.T, opt ...TestOption) (*DB, string) {
...
+			// open a new connection, which would create a new in-memory db vs
+			// using the existing one already opened in this function a few
+			// lines above with: db, err := Open(dbType, url)
+			// see: https://www.sqlite.org/inmemorydb.html
+			//
+			// luckily, the gorm sqlite ConnPool is the existing opened *sql.DB,
+			// so we can just run the migration on that conn
+			var ok bool
+			rawDB, ok = db.wrapped.ConnPool.(*sql.DB)
+			require.True(ok)
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply testing.go:125]: ty!

File: testing_test.go (hunk 1 above)
Comment (jimlambrt) [reply testing_test.go:108]: ty!

File: testing_test.go (hunk 2 above)
Comment (jimlambrt) [reply testing_test.go:66]: I went back and forth on that.... if the combination is passed and the user tries to access a db object in the schema, they'll get an error.  Given that I wasn't too worried about the user NOT knowning and it's a test option anywho.  It is helpful to the maintainer as is, so I decided to allow the combination.

**Pull request merged**